core.d.ts

```diff
@@ -0,0 +1,146 @@
+type FilterFunctionValue = Filter
+type FilterFunctions = {
+	[P in FilterFunctionKey]: FilterFunctionValue
+}
+
+interface ResourceFilter extends ResourceObj<Filter> {}
+
+type FilterObj = FilterOperations & FilterFunctions & {
+	$raw?: RawFilter
+
+	$?: string | string[]
+
+	$and?: Filter
+	$or?: Filter
+
+	$in?: Filter
+
+	$not?: Filter
+
+	$any?: Lambda
+	$all?: Lambda
+} & ResourceFilter
+
+interface FilterArray extends Array<Filter> {}
+type FilterString = string
+// Strictly speaking `[ string, ...Filter ]` but there isn't a way to type that yet
+type RawFilter = string | (string | Filter)[] | {
+	$string: string
+	[index: string]: Filter
+}
+type Lambda = {
+	$alias: string
+	$expr: Filter
+}
+type Filter = number | FilterString | FilterObj | FilterArray
+
+interface ResourceExpand extends ResourceObj<Expand> {}
+type ExpandObject = ODataOptions & ResourceExpand
+
+interface ExpandArray extends Array<Expand> {}
+type Expand = string | ExpandArray | ExpandObject
+
+type OrderBy = string | string[] | {
+	[index: string]: 'asc' | 'desc'
+}
+
+type ODataOptions = {
+	$filter?: Filter
+	$expand?: Expand
+	$orderby?: OrderBy
+	$top?: number
+	$skip?: number
+	$select?: string | string[]
+}
+type OptionsObject = ODataOptions & {
+	[index: string]: string | string[]
+}
+
+type ODataMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE'
+
+type AnyObject = {
+	[index: string]: any
+}
+
+type ResourceId = string | number | Date
+
+type SharedParam = 'apiPrefix' | 'passthrough' | 'passthroughByMethod'
+type Params = {
+	apiPrefix?: string
+	method?: ODataMethod
+	resource?: string
+	id?: ResourceId
+	url?: string
+	body?: AnyObject
+	passthrough?: AnyObject
+	passthroughByMethod?: {
+		GET: AnyObject
+		POST: AnyObject
+		PATCH: AnyObject
+		DELETE: AnyObject
+	}
+	customOptions?: AnyObject
+	options?: ODataOptions
+}
+
+export abstract class PinejsClientCore<T> {
+	apiPrefix: string
+	passthrough: AnyObject
+	passthroughByMethod: AnyObject
+	backendParams: AnyObject
+
+	// `backendParams` must be used by a backend for any additional parameters it may have.
+	constructor(params: string | Params, backendParams: AnyObject)
+
+
+	// `backendParams` must be used by a backend for any additional parameters it may have.
+	clone(params: string | Params, backendParams: AnyObject): T
+
+	query(params: Params): Promise<number | AnyObject | AnyObject[]>
+
+	get(params: Params): Promise<number | AnyObject | AnyObject[]>
+
+	put(params: Params): Promise<{}>
+
+	patch(params: Params): Promise<{}>
+
+	post(params: Params): Promise<{}>
+
+	delete(params: Params): Promise<{}>
+
+	compile(params: Params): string
+
+	request(params: Params, overrides: { method?: ODataMethod }): Promise<{}>
+
+	abstract _request: (
+		params: {
+			method: string,
+			url: string,
+			body?: AnyObject,
+		} & AnyObject) => Promise<{}>
+}
+
+
+export default (utils: Util, Promise: PromiseRejector) => typeof PinejsClientCore
```

Comment:
Afaik, you either `declare` it to put it into the global types scope, or export it (which should be fine?)